docs/devcontainerjson-reference.md

```diff
@@ -0,0 +1,172 @@
+| `build.target` | string | A string that specifies a [Docker image build target](https://docs.docker.com/engine/reference/commandline/build/#specifying-target-build-stage---target) that should be passed when building a Dockerfile. Defaults to not set. For example: `"build": { "target": "development" }` |
+| `build.cacheFrom` | string,<br>array | A string or array of strings that specify one or more images to use as caches when building the image. Cached image identifiers are passed to the `docker build` command with `--cache-from`.
+| `appPort` | integer,<br>string,<br>array |  In most cases, we recommend using the new [forwardPorts property](https://code.visualstudio.com/docs/remote/containers#_always-forwarding-a-port). This property accepts a port or array of ports that should be [published](https://code.visualstudio.com/docs/remote/containers#_publishing-a-port) locally when the container is running. Unlike `forwardPorts`, your application may need to listen on all interfaces (`0.0.0.0`) not just `localhost` for it to be available externally. Defaults to `[]`. |
+| `containerEnv` | object | A set of name-value pairs that sets or overrides environment variables for the container. Environment and [pre-defined variables](#variables-in-devcontainerjson) may be referenced in the values. For example:<br/> `"containerEnv": { "MY_VARIABLE": "${localEnv:MY_VARIABLE}" }`<br /> Requires the container be recreated / rebuilt to change. If you want to reference an existing container variable while setting this one (like updating the `PATH`), use `remoteEnv` instead. |
+| `containerUser` | string | Overrides the user for all operations run as inside the container. Defaults to either `root` or the last `USER` instruction in the related Dockerfile used to create the image.<br>On Linux, the specified container user's UID/GID will be updated to match the local user's UID/GID to avoid permission problems with bind mounts (unless disabled using `updateRemoteUserUID`).<br>Requires the container be recreated / rebuilt for updates to take effect. If you want any connected tools or related processes to use a different user than the one for the container, see `remoteUser`. |
+| `mounts` | array | An array of additional mount points to add to the container when created. Each value is a string that accepts the same values as the [Docker CLI `--mount` flag](https://docs.docker.com/engine/reference/commandline/run/#add-bind-mounts-or-volumes-using-the---mount-flag). Environment and [pre-defined variables](#variables-in-devcontainerjson) may be referenced in the value. For example:<br />`"mounts": ["source=${localWorkspaceFolder}/app-scripts,target=/usr/local/share/app-scripts,type=bind,consistency=cached"]` <br /><br />⚠️ Codespaces ignores "bind" mounts with the exception of the Docker socket. Volume mounts are still allowed.|
```

Comment:
Hmmm. Notes about Codespaces are also out of place here since its likely other tools will not fully implement the spec. There's several others throughout the doc.

@2percentsilk - Would it make sense to instead have a page in GitHub documentation listing these gaps? Or we could have a page where people can PR information on tool support into this repo?

Reply:
I added a section to `supporting-tools.md` about Codespaces-specific properties, starting out with the differences and limitations mentioned throughout the doc. Let me know if it looks reasonable to you.

I also noticed several mentions of properties `Not yet supported when using Clone Repository in Container Volume`. Should we also move those into `supporting-tools.md`?

Reply:
Yeah that's a great point.